app/models/good_job/batch.rb

```diff
@@ -86,15 +86,23 @@ def initialize(_record: nil, **properties) # rubocop:disable Lint/UnderscorePref
     # @return [Array<ActiveJob::Base>] Active jobs added to the batch
     def enqueue(active_jobs = [], **properties, &block)
       assign_properties(properties)
-      record.save!
+      if record.new_record?
+        record.save!
+      else
+        record.with_advisory_lock(function: "pg_advisory_lock") do
+          record.enqueued_at_will_change!
+          record.finished_at_will_change!
+          record.update!(enqueued_at: nil, finished_at: nil)
+        end
+      end
 
       active_jobs = add(active_jobs, &block)
 
-      record.finished_at = nil
-      record.enqueued_at = Time.current if enqueued_at.nil?
-      record.save!
+      record.with_advisory_lock(function: "pg_advisory_lock") do
+        record.update!(enqueued_at: Time.current)
+        record._continue_discard_or_finish(lock: false)
+      end
```

Comment:
Looks good! i'm agnostic about the enqueued_at reset, I think you're doing that in lieu of just putting the `add` in the advisory lock (if that's even possible)? Perfectly fine solution but i would consider enqueued_at to mean when the first jobs started coming in, doesn't matter that much though total batch statistics can be derived from the jobs.